Incremental submodule update not working after lcm repository switched

Reproduction steps:
```
git clone 'https://jwnimmer-tri@github.com/jwnimmer-tri/drake.git' drake-distro
cd drake-distro
git remote add upstream 'https://jwnimmer-tri@github.com/RobotLocomotion/drake.git'
git remote set-url --push upstream no_push
git reset --hard 8535fc31ec624e632b05c6212a7134947a59a28f   # Just before #2809.
mkdir -p build/debug
cd build/debug
env PATH=/usr/lib/ccache:$PATH CC=gcc-4.9 CXX=g++-4.9 \
  cmake ../../ -G Ninja -DCMAKE_BUILD_TYPE:STRING=Debug
env PATH=/usr/lib/ccache:$PATH CC=gcc-4.9 CXX=g++-4.9 ninja -j1 -v
git fetch upstream && git merge --ff-only upstream/master  # Takes on #2809.
env PATH=/usr/lib/ccache:$PATH CC=gcc-4.9 CXX=g++-4.9 ninja -j1 -v
```

The last build line (after merging 2809, then re-rerunning ninja), yields:
```
[27/88] cd /home/jwnimmer/jwnimmer-tri/drake-distro/externals/lcm && /home/jwnimmer/jwnimmer-tri/cmake-3.5.2-Linux-x86_64/bin/cmake -E chdir /home/jwnimmer/jwnimmer-tri/drake-distro /usr/bin/git submodule update --init --recursive -- externals/lcm && /home/jwnimmer/jwnimmer-tri/cmake-3.5.2-Linux-x86_64/bin/cmake -E touch /home/jwnimmer/jwnimmer-tri/drake-distro/build/debug/lcm-prefix/src/lcm-stamp/lcm-update
FAILED: cd /home/jwnimmer/jwnimmer-tri/drake-distro/externals/lcm && /home/jwnimmer/jwnimmer-tri/cmake-3.5.2-Linux-x86_64/bin/cmake -E chdir /home/jwnimmer/jwnimmer-tri/drake-distro /usr/bin/git submodule update --init --recursive -- externals/lcm && /home/jwnimmer/jwnimmer-tri/cmake-3.5.2-Linux-x86_64/bin/cmake -E touch /home/jwnimmer/jwnimmer-tri/drake-distro/build/debug/lcm-prefix/src/lcm-stamp/lcm-update
fatal: reference is not a tree: 9c9e22cf23663ef9cd0e5ec83932ff7f70cd7b18
Unable to checkout '9c9e22cf23663ef9cd0e5ec83932ff7f70cd7b18' in submodule path 'externals/lcm'
ninja: build stopped: subcommand failed.
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -54,6 +54,18 @@ function(drake_forceupdate proj)
   endif()
 endfunction()
 
+function(drake_depend_submodule_sync proj)
+  # All targets that contain a submodule update command, or depend on a
+  # target that does, need to depend on the submodule-sync target.
+  if(AUTO_UPDATE_EXTERNALS)
+    foreach(t ${proj} ${proj}-update)
+      if(TARGET ${t})
+        add_dependencies(${t} submodule-sync)
+      endif()
+    endforeach()
+  endif()
+endfunction()
+
 # Require Java at the super-build level since libbot cannot configure without finding Java
 find_package(Java 1.6 REQUIRED)
 include(UseJava)
@@ -339,10 +351,6 @@ foreach(proj ${EXTERNAL_PROJECTS})
 
     if(AUTO_UPDATE_EXTERNALS)
       set(${proj}_UPDATE_COMMAND ${CMAKE_COMMAND} -E chdir ${CMAKE_CURRENT_SOURCE_DIR} ${${proj}_DOWNLOAD_COMMAND})
-
-      # Synchronize the submodule url now so parallel updates do not conflict later.
-      execute_process(COMMAND ${GIT_EXECUTABLE} submodule sync -- ${${proj}_GIT_SUBMODULE_PATH}
-        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
     else()
       set(${proj}_UPDATE_COMMAND "")
     endif()
@@ -378,6 +386,7 @@ foreach(proj ${EXTERNAL_PROJECTS})
         ${${proj}_ADDITIONAL_CMAKE_CONFIGURE_ARGS}
         ${${proj}_ADDITIONAL_BUILD_ARGS})
       drake_forceupdate(${proj})
+      drake_depend_submodule_sync(${proj})
     else() # not a CMake POD
       if(NOT ${proj}_BINARY_DIR)
         # In-source build for non-CMake projects.
@@ -418,6 +427,7 @@ foreach(proj ${EXTERNAL_PROJECTS})
         DEPENDS ${deps}
         ${${proj}_ADDITIONAL_BUILD_ARGS})
       drake_forceupdate(${proj})
+      drake_depend_submodule_sync(${proj})
       # message(STATUS "${proj}_BUILD_COMMAND: ${${proj}_BUILD_COMMAND}")
     endif()
   else()
@@ -442,6 +452,7 @@ foreach(proj ${EXTERNAL_PROJECTS})
       ${PODS_VERBOSE_MAKEFILE}
       ${${proj}_ADDITIONAL_CMAKE_CONFIGURE_ARGS}
       ${${proj}_ADDITIONAL_BUILD_ARGS})
+    drake_depend_submodule_sync(${proj})
   endif()
 
 endforeach()
@@ -502,10 +513,6 @@ foreach(example IN ITEMS LittleDog)
     set(${example}_DOWNLOAD_COMMAND ${GIT_EXECUTABLE} submodule update --init --recursive -- drake/examples/${example})
     if(AUTO_UPDATE_EXTERNALS)
       set(${example}_UPDATE_COMMAND ${CMAKE_COMMAND} -E chdir ${CMAKE_CURRENT_SOURCE_DIR} ${${example}_DOWNLOAD_COMMAND})
-
-      # Synchronize the submodule url now so parallel updates do not conflict later.
-      execute_process(COMMAND ${GIT_EXECUTABLE} submodule sync -- drake/examples/${example}
-        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
     else()
       set(${example}_UPDATE_COMMAND "")
     endif()
@@ -519,10 +526,20 @@ foreach(example IN ITEMS LittleDog)
       BUILD_COMMAND ""
       INSTALL_COMMAND "")
     drake_forceupdate(download-${example})
+    drake_depend_submodule_sync(download-${example})
     add_dependencies(drake download-${example})
     add_dependencies(download-all download-${example})
   endif()
 endforeach()
+
+if(AUTO_UPDATE_EXTERNALS)
+  # Start every build by synchronizing submodule URLs.  We do this as a
+  # separate target so all URLs can be synchronized at once instead of
+  # conflicting during parallel updates later.
+  add_custom_target(submodule-sync
+    COMMAND ${GIT_EXECUTABLE} submodule --quiet sync --
+    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
+endif()
 
 string(REPLACE ";" " " PROJECT_LIST "${PROJECT_LIST}")
 add_custom_target(list-project-dirs COMMAND echo "${PROJECT_LIST}")

